### PR TITLE
Added IIncludeContext, implemented missing properties, added includeNdlc, sync NDLC and NDC

### PR DESCRIPTION
**Pull request by 304NotModified**

### CODE REVIEW

File: src/NLog/LayoutRenderers/Log4JXmlEventLayoutRenderer.cs [hunk 1]
```diff
@@ -233,6 +247,10 @@ protected override void Append(StringBuilder builder, LogEventInfo logEvent)
                 {
                     xtw.WriteElementSafeString("log4j", "NDC", dummyNamespace, string.Join(this.NdcItemSeparator, NestedDiagnosticsContext.GetAllMessages()));
                 }
+                if (this.IncludeNdlc)
+                {
+                    xtw.WriteElementSafeString("log4j", "NDLC", dummyNamespace, string.Join(this.NdcItemSeparator, NestedDiagnosticsLogicalContext.GetAllObjects()));
```
Comment (304NotModified) [inline src/NLog/LayoutRenderers/Log4JXmlEventLayoutRenderer.cs]: todo we don't have a `GetAllMessages` ?

File: src/NLog/Layouts/JsonLayout.cs [hunk 2]
```diff
@@ -92,6 +94,18 @@ public JsonLayout()
...
+        /// Gets or sets a value indicating whether to include contents of the <see cref="NestedDiagnosticsContext"/> stack.
+        /// </summary>
+        /// <docgen category='Payload Options' order='10' />
+        public bool IncludeNdc { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether to include contents of the <see cref="NestedDiagnosticsLogicalContext"/> stack.
+        /// </summary>
+        /// <docgen category='Payload Options' order='10' />
+        public bool IncludeNdlc { get; set; }
```
Comment (304NotModified) [inline src/NLog/Layouts/JsonLayout.cs]: todo useful for JsonLayout?

File: src/NLog/Layouts/JsonLayout.cs [hunk 3]
```diff
@@ -92,6 +94,18 @@ public JsonLayout()
         public bool IncludeAllProperties { get; set; }
 
         /// <summary>
+        /// Gets or sets a value indicating whether to include contents of the <see cref="NestedDiagnosticsContext"/> stack.
+        /// </summary>
+        /// <docgen category='Payload Options' order='10' />
+        public bool IncludeNdc { get; set; }
```
Comment (304NotModified) [inline src/NLog/Layouts/JsonLayout.cs]: todo useful for JsonLayout? How to implement? (JSON list?)

File: src/NLog/LayoutRenderers/Log4JXmlEventLayoutRenderer.cs (hunk 1 above)
Comment (snakefoot) [inline src/NLog/LayoutRenderers/Log4JXmlEventLayoutRenderer.cs]: Nothing is stopping you from adding it :)

**Comment by snakefoot** [comment]
@304NotModified Actually tried to google how structured logging preferred to use context-details received from Microsoft-ILogger-BeginScope when outputing the json for elastic search. But I failed in finding any good answers :(

I guess one would consider the scope-details as meta-data-list or as tags-list. So the rendering of the NDC / NDLC should probably be outputted by a LayoutRenderer that can be referenced by the JsonLayout as a named-list-property (Allow user-defined name for this list).

File: src/NLog/LayoutRenderers/Log4JXmlEventLayoutRenderer.cs (hunk 1 above)
Comment (304NotModified) [inline src/NLog/LayoutRenderers/Log4JXmlEventLayoutRenderer.cs]: done :)

**Comment by snakefoot** [comment]
Maybe extract IIncludeProperties from IIncludeContext, and change JsonLayout to inherit only from this.

**Comment by 304NotModified** [comment]
yes also an option, but I think it's nice that the jsonlayout also could handle NDC/NDLC. Or is that useless in your opinion?

**Comment by snakefoot** [comment]
Well I like that the JSON-Layout is completely free of hardcoded json-member-names. Outputting the NDC / NDLC will require using a property-name.

I guess you could change that the IncludeNdc and IncludeNdlc into a string-type, and when not-null-or-empty, then it will actually output them as array using the member-name-specified.

**Comment by 304NotModified** [comment]
> Well I like that the JSON-Layout is completely free of hardcoded json-property-names. Outputting the NDC / NDLC will require using a property-name.

Thanks for the feedback. Have removed the JsonLayout changes in this PR. Maybe in the future

**Comment by 304NotModified** [comment]
@snakefoot any idea if there are specs which are valid log4j xml elements?

**Comment by snakefoot** [comment]
In log4j the NDLC is translated into the name "Thread Context" and MDLC is translated into "Thread Context Map". And normally they values are stored as tags-lists when saved into log-frameworks.

Found this code:

https://logging.apache.org/log4j/1.2/xref/org/apache/log4j/xml/XMLLayout.html

And it seems that NDLC and NDC should be put into the same xml-tag (NDC).

And it seems that MDC, MDLC and LogEventInfo.Properties should be put into the same xml-tag (Properties).

**Comment by 304NotModified** [comment]
So in summary I only have to fix the ndlc? Do you think we could have 2 times a NDC element? (For NDC and ndlc)

**Comment by snakefoot** [comment]
> Do you think we could have 2 times a NDC element? (For NDC and ndlc)

No they should be put into the same NDC-xml-tag.

**Comment by 304NotModified** [comment]
That would be weird when both are used, isn't?

**Comment by snakefoot** [comment]
> That would be weird when both are used, isn't?

Then you can also change the IncludeMDC and IncludeMDLC into a breaking change, and have multiple Properties-tags.

**Comment by 304NotModified** [comment]
```
.................................................................................................................................................................................................................................System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.Runtime.Serialization.SerializationException: Type 'NLog.NestedDiagnosticsLogicalContext+NestedContext`1[[System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]' in Assembly 'NLog, Version=0.0.0.0, Culture=neutral, PublicKeyToken=5120e14c03d0593c' is not marked as serializable.
Server stack trace: 
   at System.Runtime.Serialization.FormatterServices.InternalGetSerializableMembers(RuntimeType type)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at System.Runtime.Serialization.FormatterServices.GetSerializableMembers(Type type, StreamingContext context)
   at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.InitMemberInfo()
   at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.InitSerialize(Object obj, ISurrogateSelector surrogateSelector, StreamingContext context, SerObjectInfoInit serObjectInfoInit, IFormatterConverter converter, ObjectWriter objectWriter, SerializationBinder binder)
   at System.Runtime.Serialization.Formatters.Binary.ObjectWriter.Write(WriteObjectInfo objectInfo, NameInfo memberNameInfo, NameInfo typeNameInfo)
   at System.Runtime.Serialization.Formatters.Binary.ObjectWriter.Serialize(Object graph, Header[] inHeaders, __BinaryWriter serWriter, Boolean fCheck)
   at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize(Stream serializationStream, Object graph, Header[] headers, Boolean fCheck)
   at System.Runtime.Remoting.Channels.CrossAppDomainSerializer.SerializeMessageParts(ArrayList argsToSerialize)
   at System.Runtime.Remoting.Messaging.SmuggledMethodCallMessage..ctor(IMethodCallMessage mcm)
   at System.Runtime.Remoting.Messaging.SmuggledMethodCallMessage.SmuggleIfPossible(IMessage msg)
   at System.Runtime.Remoting.Channels.CrossAppDomainSink.SyncProcessMessage(IMessage reqMsg)
Exception rethrown at [0]: 
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at System.Web.UI.ICallbackEventHandler.RaiseCallbackEvent(String eventArgument)
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Xunit.Sdk.ExecutorCallback.CallbackEventHandlerCallback.Notify(String value)
   at Xunit.Sdk.Executor.OnTestResult(ITestResult result, ExecutorCallback callback)
   at Xunit.Sdk.TestClassCommandRunner.Execute(ITestClassCommand testClassCommand, List`1 methods, Predicate`1 startCallback, Predicate`1 resultCallback)
   at Xunit.Sdk.Executor.ThreadRunner(Object threadStart)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart(Object obj)
Command exited with code 1
```

**Comment by snakefoot** [comment]
@304NotModified This is a breaking change in NestedDiagnosticsLogicalContext:

> 
> In assembly C:\Users\snakefoot\Source\Repos\BitDiffer_1_5_0_4\4.4.12\net45\NLog.dll:
> 
> public static object Pop()
> 
> In assembly C:\Users\snakefoot\Source\Repos\BitDiffer_1_5_0_4\4.5\net45\NLog.dll:
> 
> public static string Pop()
> 
> Changes Found : Implementation changed, Declaration changed (Breaking)

But I guess this has been considered during implementation.

**Comment by 304NotModified** [comment]
Nope,  it's a bug!

**Comment by snakefoot** [comment]
> Nope, it's a bug!

Think I would let this bug slip through, since it ensures that NDC and NDLC matches the same "invisible" interface. Espcially since NDLC is a new kid on the block, so the chance that people have started using Pop() directly is very low, since Push provides a much nicer IDisposable that performs implicit Pop.

**Comment by snakefoot** [comment]
@304NotModified Updated documentation:

https://github.com/NLog/NLog/wiki/NLogViewer-target (Added IncludeNdlc + NdlcItemSeparator + IncludeAllProperties)
https://github.com/NLog/NLog/wiki/Log4JXmlEventLayout (Added IncludeNdlc)
https://github.com/NLog/NLog/wiki/Log4JXMLEvent-Layout-Renderer (Added IncludeNdlc + NdlcItemSeparator)

**Pull request merged**